odl/discr/partition.py

```diff
@@ -766,6 +767,132 @@ def uniform_partition_fromgrid(grid, begin=None, end=None):
     return RectPartition(IntervalProd(begin_vec, end_vec), grid)
 
 
+def uniform_partition_options(begin=None, end=None, num_nodes=None,
+                              cell_sides=None):
+    """Return a partition with equally sized cells.
+
+    This function offers more flexibility over `uniform_discr` in that
+    any combination of 3 or 4 out of 4 possible init parameters can be
+    given, and the (possibly) missing one is calculated.
+
+    Parameters
+    ----------
+    begin, end : float or array-like, optional
+        Vectors defining the begin end end points of an `IntervalProd`
+        (a rectangular box). For one-dimensional partitions, single
+        floats can be provided. ``None`` entries mean "compute the value".
+    num_nodes : int or sequence of int, optional
+        Number of nodes per axis. For 1d intervals, a single integer
+        can be specified. ``None`` entries mean "compute the value".
+    cell_sides : float or array-like, optional
+        Side length of the partition cells per axis. For 1d intervals,
+        a single integer can be specified. ``None`` entries mean
+        "compute the value".
+
```

Comment:
I thought it would become too complicated. But I'll go through the different situations and check what we got. We have the following 4 cases for provided parameters:

- `begin`, `end`, `num_nodes`: equivalent to `uniform_sampling`, easy to introduce `nodes_on_bdry`
- `begin`, `end`, `cell_sides`: Here, `nodes_on_bdry` could be used to determine if we have half cells only at the boundaries. That would mean to calculate `num_nodes` differently.
- `begin`, `num_nodes`, `cell_sides`: Same as the second case, but here the computation of `end` has to be adapted.
- `end`, `num_nodes`, `cell_sides`: Backwards version of the third case

Conclusion: ok, why not?